### PR TITLE
Additional lints

### PR DESCRIPTION
**Pull request by TimWhiting**
Working on some additional lints:
- [x] Finish riverpod_no_mutate_sync  (don't mutate providers synchronously inside build of widget or providers).
- [x] Avoid read on autoDispose / use container.listen on autoDispose
- [x] use ref synchronously in widget callbacks (similar to BuildContext across async gaps)
- [x] Avoid global provider container
- [x] Finish riverpod_ref_escape_scope (don't pass WidgetRef between widgets, dont pass Ref except to a Notifier constructor or field initializer

### CODE REVIEW

**Comment by rrousselGit** [comment]
Epic!

**Comment by rrousselGit** [comment]
By the way, it'd be nice if the lints could be made compatible with the code-generator too.

Some of them should be fine, but I'm sure some will fail

**Comment by TimWhiting** [comment]
> By the way, it'd be nice if the lints could be made compatible with the code-generator too.
> 
> Some of them should be fine, but I'm sure some will fail

Yes, I was thinking the same thing. I was playing with the code generator for the first time yesterday.

**Comment by TimWhiting** [comment]
I had to increase the timeout on the test. In CI it might be even slower. Is the slowness because of the custom_lint package, or because of how the lint rules are written do you think? There are probably some caching / other optimizations I should do with some of the lints, but I just wanted to know if it is partly the custom_lint package?

**Comment by rrousselGit** [comment]
About riverpod_no_mutate_sync, a case I tought of is `ref.invalidate`/`ref.invalidateSelf`

They probably should not be allowed either, using the same reasoning.

**Comment by TimWhiting** [comment]
> About riverpod_no_mutate_sync, a case I tought of is `ref.invalidate`/`ref.invalidateSelf`
> 
> They probably should not be allowed either, using the same reasoning.

What about `ref.refresh`?

**Comment by TimWhiting** [comment]
Let's merge this as is. I think this is getting too big for one PR. I've added items on the Linter issue for improving performance, and adding support for the codegen.

**Comment by TimWhiting** [comment]
I'll try to fix the build before merging though

**Comment by TimWhiting** [comment]
@rrousselGit Ready for review. Compatibility with codegen, and figuring out how to speed up / optimize these lints will be next.

File: packages/riverpod_lint_flutter_test/test/flutter_test.dart [hunk 1]
```diff
@@ -54,8 +79,14 @@ Future<void> main() async {
...
+  test/goldens/use_ref_before_async_gaps.dart:48:11 • Do not use ref after async gaps in flutter widgets. • riverpod_no_ref_after_async
+  test/goldens/use_ref_before_async_gaps.dart:51:9 • Do not use ref after async gaps in flutter widgets, a function was called which uses ref after a widget could be disposed • riverpod_no_ref_after_async
+  test/goldens/use_ref_before_async_gaps.dart:53:9 • Do not use ref after async gaps in flutter widgets, a function was called which uses ref after a widget could be disposed • riverpod_no_ref_after_async
+  test/goldens/use_ref_before_async_gaps.dart:55:9 • Do not use ref after async gaps in flutter widgets, a function was called which uses ref after a widget could be disposed • riverpod_no_ref_after_async
+  test/goldens/use_ref_before_async_gaps.dart:57:15 • Do not use ref after async gaps in flutter widgets, a function was called which uses ref after a widget could be disposed • riverpod_no_ref_after_async
+  test/goldens/use_ref_before_async_gaps.dart:59:9 • Do not use ref after async gaps in flutter widgets. • riverpod_no_ref_after_async
 ''');
-  }, skip: 'TODO must fix');
+  });
```
Comment (rrousselGit) [inline packages/riverpod_lint_flutter_test/test/flutter_test.dart:91]:
```suggestion
  }, skip: 'TODO flacky');
```

**Comment by rrousselGit** [comment]
Awesome thanks!

**Pull request merged**